erpnext/accounts/doctype/pos_invoice/pos_invoice.py

```diff
@@ -740,3 +740,7 @@ def append_payment(payment_mode):
 		]:
 			payment_mode = get_mode_of_payment_info(mode_of_payment, doc.company)
 			append_payment(payment_mode[0])
+
+
+def on_doctype_update():
+	frappe.db.add_index("POS Invoice", ["customer", "is_return", "return_against"])
```

Comment:
Just a simple return against index should be enough. It has high enough cardinality ~= unique names. 

That can be managed from doctype directly 👀